Fix impersonate issue caused due to core change

sharidas opened the pull request:
Due to core change in user session, impersonation
was failing to happen. This change resolves the issue.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
Have to update the tests.

PVince81 commented on appinfo/app.php:
```diff
@@ -32,7 +32,9 @@ function() {
 	OC::$server->getLogger(),
-	OC::$server->getSession()
+	OC::$server->getSession(),
+	\OC::$server->query('\OC\Authentication\Token\DefaultTokenProvider'),
```
in the future let's move all this into a proper Application class where the services are registered

PVince81 commented on controller/logoutcontroller.php:
```diff
@@ -40,13 +44,17 @@ class LogoutController extends Controller {
-    public function __construct($appName, IRequest $request,IUserManager $userManager, IUserSession $userSession, ILogger $logger, ISession $session) {
+    public function __construct($appName, IRequest $request,IUserManager $userManager,
+								IUserSession $userSession, ILogger $logger, ISession $session,
+								DefaultTokenProvider $tokenProvider, Session $ocUserSession) {
```
always use interfaces, in this case \OC\Authentication\Token\IProvider

PVince81 commented on controller/logoutcontroller.php:
```diff
@@ -71,7 +79,10 @@ public function logoutcontroller(GenericEvent $event) {
+			$sessionId = $this->session->getId();
+			$this->tokenProvider->invalidateToken($sessionId);
 			$this->userSession->setUser($impersonatorUser);
+			$this->ocUserSession->createSessionToken($this->request, $impersonatorUser->getUID(), $impersonatorUser->getDisplayName());
```
Login name is not the display name. In this one case I suggest using the UID. I hope this won't break other things.

@DeepDiver1975 @butonic

PVince81 commented on controller/settingscontroller.php:
```diff
@@ -54,10 +59,17 @@ class SettingsController extends Controller {
 				IUserSession $userSession, ILogger $logger, IGroupManager $groupManager,
-				SubAdmin $subAdmin, ISession $session, IAppConfig $config, IL10N $l10n) {
+				SubAdmin $subAdmin, ISession $session, IAppConfig $config, IL10N $l10n,
+				DefaultTokenProvider $tokenProvider) {
```
use interface

PVince81 commented on controller/settingscontroller.php:
```diff
@@ -131,7 +145,13 @@ public function impersonate($target) {
+				$sessionId = $this->session->getId();
+				$this->tokenProvider->invalidateToken($sessionId);
 				$this->userSession->setUser($user);
+				$this->ocUserSession->createSessionToken($this->request, $user->getUID(), $user->getDisplayName());
```
same comment about login name

PVince81 commented on controller/settingscontroller.php:
```diff
@@ -142,7 +162,13 @@ public function impersonate($target) {
 				foreach ($includedGroups as $group) {
 					if($this->subAdmin->isSubAdminofGroup($this->userSession->getUser(), $this->groupManager->get($group))) {
 						$this->logger->info("User $impersonator impersonated user $target", ['app' => 'impersonate']);
+						$sessionId = $this->session->getId();
```
can we refactor this logic into a utility method ? the logic about session id, invalidate token, set user, create session token is repeated in three locations

PVince81 commented on appinfo/app.php:
```diff
@@ -32,7 +32,9 @@ function() {
-	OC::$server->getSession()
+	OC::$server->getSession(),
+	\OC::$server->query('\OC\Authentication\Token\DefaultTokenProvider'),
+	\OC::$server->getUserSession()
```
there is already a getUserSession above, so you have twice the same arg

PVince81 commented:
Tested, works so far. Let's retest after refactoring.

sharidas commented:
Updated the code change, as requested. Some changes couldn't make it:
- DefaultTokenProvider to \OC\Authentication\Token\IProvider. When change made it was causing trouble and exception. So kept it like that.
```{"reqId":"ncRvuSfQBTAyNSSMJgue","level":3,"time":"2017-11-14T09:02:02+00:00","remoteAddr":"127.0.0.1","user":"admin","app":"index","method":"GET","url":"\/testing\/index.php\/apps\/impersonate\/getimpersonatedata","message":"Exception: {\"Exception\":\"OCP\\\\AppFramework\\\\QueryException\",\"Message\":\"Could not resolve OC\\\\Authentication\\\\Token\\\\IProvider! Class can not be instantiated\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(113): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('OC\\\\\\\\Authenticati...')\\n#1 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(65): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('OC\\\\\\\\Authenticati...')\\n#2 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(92): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->buildClass(Object(ReflectionClass))\\n#3 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php(113): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->resolve('OCA\\\\\\\\Impersonate...')\\n#4 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/App.php(85): OC\\\\AppFramework\\\\Utility\\\\SimpleContainer->query('OCA\\\\\\\\Impersonate...')\\n#5 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Routing\\\/RouteActionHandler.php(46): OC\\\\AppFramework\\\\App::main('OCA\\\\\\\\Impersonate...', 'getDataForImper...', Object(OC\\\\AppFramework\\\\DependencyInjection\\\\DIContainer), Array)\\n#6 [internal function]: OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler->__invoke(Array)\\n#7 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/Route\\\/Router.php(342): call_user_func(Object(OC\\\\AppFramework\\\\Routing\\\\RouteActionHandler), Array)\\n#8 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/base.php(931): OC\\\\Route\\\\Router->match('\\\/apps\\\/impersona...')\\n#9 \\\/home\\\/sujith\\\/test\\\/owncloud\\\/index.php(56): OC::handleRequest()\\n#10 {main}\",\"File\":\"\\\/home\\\/sujith\\\/test\\\/owncloud\\\/lib\\\/private\\\/AppFramework\\\/Utility\\\/SimpleContainer.php\",\"Line\":94}"}```

The changes made:
- Replaced displayname from being used in the createSessionToken param with `UID`.
- Created a Util class with a single function switchUser. This method is called by both logincontroller and settingscontroller.

PVince81 commented on lib/Util.php:
```diff
@@ -0,0 +1,46 @@
+	 * @param \OC\User\Session $ocUserSession
+	 * @param \OCP\IRequest $request
+	 */
+	public function switchUser(\OCP\ISession $session,
```
I suggest injecting these into the constructor instead. So `switchUser` only takes the target user.

Is there a chance to also add the logic about keeping track of the impersonating user here, if that makes sense ?

sharidas commented:
- Added constructor for Util and did the injection over there
- Modified the `switchUser()` to accept user to be switched and the impersonator. Also added tracking of impersonated user in `switchUser()`

PVince81 commented on lib/Util.php:
```diff
@@ -0,0 +1,54 @@
+use OCP\IUserSession;
+
+class Util {
+	private $session;
```
missing PHPDoc with `@var` specifying the types

PVince81 commented on controller/logoutcontroller.php:
```diff
@@ -71,7 +81,7 @@ public function logoutcontroller(GenericEvent $event) {
 			], Http::STATUS_NOT_FOUND);
 		} else {
-			$this->userSession->setUser($impersonatorUser);
+			$this->util->switchUser($impersonatorUser, $this->session->get('impersonator'));
```
shouldn't we clear the impersonator with `null` in this case on logout ?

sharidas commented:
- Updated Util with PHPDoc
- Verified with logger if the session variable is getting removed when logout operation is performed. For example tried adding a log at the https://github.com/owncloud/impersonate/pull/80/files#diff-bbd3446ee19e6a7cee3fcca374725eb6L79. So when the admin/subAdmin navigates to the users page. The log file populates, and I verified that initially the session variable is empty.
  - Next tried to add logger around the `switchUser()` in util. Basically at the end to check if after logout the variable still exists or not. And at the beginning to check if the variable is set or not.
  - When impersonated successfully, the variable is set ( the log file data : ```{"reqId":"W0tMFxsT0hyVpJxR4EiU","level":2,"time":"2017-11-14T13:24:38+00:00","remoteAddr":"127.0.0.1","user":"user3","app":"no app in context","method":"POST","url":"\/testing\/index.php\/apps\/impersonate\/user","message":"OCA\\Impersonate\\Util::switchUser impersonator session val = user1"}
{"reqId":"W0tMFxsT0hyVpJxR4EiU","level":2,"time":"2017-11-14T13:24:38+00:00","remoteAddr":"127.0.0.1","user":"user3","app":"no app in context","method":"POST","url":"\/testing\/index.php\/apps\/impersonate\/user","message":"OCA\\Impersonate\\Util::switchUser impersonator session val after removal = user1"}```
   - When logged out successfully, the log file had data: ```{"reqId":"ZwJ8647mOZCsapbp8hTs","level":2,"time":"2017-11-14T13:24:52+00:00","remoteAddr":"127.0.0.1","user":"user1","app":"no app in context","method":"GET","url":"\/testing\/index.php\/logout?requesttoken=Nh8YdD5xRQRfAldkHz8EAA5vAggENwFRIxkiAj9kKQI%3D%3AgRvBPBpc9F60FjpHaZH1axq9mAhtJ6oQh16sSL2rEtM%3D","message":"OCA\\Impersonate\\Util::switchUser impersonator session val = user1"}
{"reqId":"ZwJ8647mOZCsapbp8hTs","level":2,"time":"2017-11-14T13:24:52+00:00","remoteAddr":"127.0.0.1","user":"user1","app":"no app in context","method":"GET","url":"\/testing\/index.php\/logout?requesttoken=Nh8YdD5xRQRfAldkHz8EAA5vAggENwFRIxkiAj9kKQI%3D%3AgRvBPBpc9F60FjpHaZH1axq9mAhtJ6oQh16sSL2rEtM%3D","message":"OCA\\Impersonate\\Util::switchUser impersonator session val after removal = "}```

PVince81 commented on lib/Util.php:
```diff
@@ -0,0 +1,79 @@
+	private $session;
+	/** @var DefaultTokenProvider  */
+	private $tokenProvider;
+	/** @var Session  */
```
isn't that IUserSession ? always use interfaces where possible

sharidas replied:
@PVince81  Ok. My intention here was if we have IUserSession, then again we need `\OC\User\Session` to call `createSessionToken()`. So I thought why not have just `\OC\User\Session` instead of IUserSession. IUserSession doesn't have declaration of `createSessionToken()` in it. I know it sounds weird. But then, please let me know your opinion in this situation.

PVince81 replied:
Hmm, in general we try to only ever inject interfaces. It's bad that the method isn't on the interface.

Ah well, then leave it as is.
